Tools/cases_generator/analyzer.py

```diff
@@ -416,11 +416,12 @@ def analyze_caches(inputs: list[parser.InputEffect]) -> list[CacheEntry]:
     caches: list[parser.CacheEffect] = [
         i for i in inputs if isinstance(i, parser.CacheEffect)
     ]
-    for cache in caches:
-        if cache.name == "unused":
-            raise analysis_error(
-                "Unused cache entry in op. Move to enclosing macro.", cache.tokens[0]
-            )
+    if caches:
+        for cache in (caches[0], caches[-1]):
+            if cache.name == "unused":
+                raise analysis_error(
+                    "First cache entry in op is unused. Move to enclosing macro.", cache.tokens[0]
+                )
```

Comment:
The `caches[-1]` here makes it so that the reader has to think about it a little more than necessary,  Consider something more explicit, for clarity:

```suggestion
    assert len(caches) <= 2, len(caches)        
    for cache in caches:
        if cache.name == "unused":
            raise analysis_error(
                "First cache entry in op is unused. Move to enclosing macro.", cache.tokens[0]
            )
```

Reply:
Also, "First cache entry" is a little confusing; as a reader without much background here, I think of each item in `cache` as a cache entry, so I'd expect the message to be something like "cache entry {index} in op is unused...".

Reply:
Your change looks good.